src/common/lock.c

```diff
@@ -280,3 +287,58 @@ lockRelease(bool failOnNoLock)
 
     FUNCTION_LOG_RETURN(BOOL, result);
 }
+
+/**********************************************************************************************************************************/
+bool
+writeLockPercentageComplete(const String *lockPath, const String *stanza, const String *execId, const double percentageComplete)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+    FUNCTION_LOG_PARAM(STRING, stanza);
+    FUNCTION_LOG_PARAM(STRING, execId);
+    FUNCTION_LOG_PARAM(DOUBLE, percentageComplete);
+    FUNCTION_LOG_END();
+
+    ASSERT(lockPath != NULL);
+    ASSERT(stanza != NULL);
+    ASSERT(execId != NULL);
+
+    bool status = false;
+
+    // This context should already exist, noop if it doesn't
+    if (lockMemContext != NULL)
+    {
+        MEM_CONTEXT_BEGIN(lockMemContext)
+        {
+            // We are only interested in the backup lock file
+            const String *const lockFPtr = strNewFmt(
+                    "%s/%s-%s" LOCK_FILE_EXT, strZ(lockPath), strZ(stanza), lockTypeName[lockTypeBackup]);
+
+            // If backup lock file exists, locate the file descriptor and update the content of the file.
+            if (storageExistsP(storageLocal(), lockFPtr))
+            {
+                LockType lockMin = lockTypeHeld == lockTypeAll ? lockTypeArchive : lockTypeHeld;
+                LockType lockMax = lockTypeHeld == lockTypeAll ? (lockTypeAll - 1) : lockTypeHeld;
+                for (LockType lockIdx = lockMin; lockIdx <= lockMax; lockIdx++)
+                {
+                    if (strEq(lockFile[lockIdx], lockFPtr))
+                    {
+                        // Build kv for second line json
+                        KeyValue *keyValue = kvNew();
+                        kvPut(keyValue, EXEC_ID_VAR, varNewStr(execId));
+                        const String *const percentageCompleteStr = strNewFmt("%.2lf", percentageComplete);
+                        kvPut(keyValue, PERCENTAGE_COMPLETE_VAR, varNewStr(percentageCompleteStr));
+
+                        // Move to the beginning of the file, and overwrite the contents with the udpated data
+                        lseek(lockFd[lockIdx], 0, SEEK_SET);
+                        ioFdWriteOneStr(lockFd[lockIdx], strNewFmt("%d" LF_Z "%s" LF_Z, getpid(), strZ(jsonFromKv(keyValue))));
```

Comment:
Check for error here using THROW_ON_SYS_ERROR. There's an example of how to do this with lseek in the Posix driver.

Reply:
updated in 0f56ed01..1f3fbd2b